### ISSUE TITLE
Support multiple users

### ISSUE THREAD

**Issue by nickavem** [cockpit-project/cockpit-podman#692]
It would be nice if the main "admin" user (1000) would be able to view the containers, images etc. of other users. I (and I'm sure others) have rootless containers on a user by user basis for maximum security, especially when dealing with volumes, etc. Right now UID 1000 can only modify the root users and their own podman installations, but it would be nice if we could modify everyone's.

**Comment by marusak** [comment]
That seems like a very bad pattern that we would give user option to see anyone's containers and manage them.
The cleanest solution is to log in as given user to manage those user containers.
We offer rootless for users in wheel group because they are somewhat special, but we are not going to do this between any users.

**Comment by nickavem** [comment]
I figured, but worth a try. Thanks!

**Comment by mailinglists35** [comment]
can you rethink this decision? I am the only user but for extra isolation I am running each container in it's own linux username. I cannot even login using those usernames because they have the login disabled for security reasons.

**Comment by martinpitt** [comment]
But this is tricky to do *precisely* because of the security isolation that you put up. These container users would each have to get a permanently running logind session ("lingering") to be able to auto-start containers and start the podman API, and yet it's difficult to log in as them. Something like `su - username` gets you the privileges, but not the `$XDG_RUNTIME_DIR` context, etc. All of this can be overcome, but you'd end up with a design which is very complex, immensely wasteful, brittle, and not even more secure as anything that can simultaneously talk to all of the containers is also the point of failure.

So I'm afraid this stands -- we can't support this, sorry.

**Comment by mailinglists35** [comment]
> But this is tricky to do _precisely_ because of the security isolation that you put up. These container users would each have to get a permanently running logind session ("lingering") to be able to auto-start containers and start the podman API, and yet it's difficult to log in as them.

I have enabled lingering for these users, ofcourse.

>Something like `su - username` gets you the privileges, but not the `$XDG_RUNTIME_DIR` context, etc. 

there is already an XDG_RUNTIME_DIR-compliant method recommended by @rhatdan from podman: [machinectl](https://lists.podman.io/archives/list/podman@lists.podman.io/thread/HGVY3CCQGHR3KHEHHSGTP6ACMJ26KO5B/) 

see also [redhat's article](https://www.redhat.com/sysadmin/sudo-rootless-podman) endorsing `machinectl`

>All of this can be overcome, but you'd end up with a design which is very complex, immensely wasteful, brittle, and not even more secure as anything that can simultaneously talk to all of the containers is also the point of failure.
> 
> So I'm afraid this stands -- we can't support this, sorry.

I would happy even with just basic read-only-effect runs of podman, like listing containers/logs, images and volumes. seeing the status of containers in cockpit would be enormously helpful to my support staff instead of giving them shell access to a sudo account just to run `machinectl`.

what does it make so complicated? it's just this:

instead of 

`podman ps -a --noheading --no-trunc` as cockpit logged on user, 

it would be

````
$ getent passwd | grep bash | awk -F: '{print $1}' | grep -v ^root$ | sort | \
xargs -Iusername \
sudo machinectl -q shell username@ \
/bin/bash -c 'podman ps -a --noheading --no-trunc'
````

**Comment by mailinglists35** [comment]
>getent passwd | grep bash 

or whatever preffered method you have to list standard users

**Comment by mailinglists35** [comment]
>anything that can simultaneously talk to all of the containers is also the point of failure

ELI5? how is this a point of failure? you still have the ssh access to the machine.

**Comment by martinpitt** [comment]
That's what I meant with "immensely wasteful/expensive" :grin: Your command might work for a quick local hack for *your specific setup*, but there's about a zillion things which are wrong with it. Plus, it essentially amounts to a DoS if you have more than a just a handful of users.

Also, c-podman uses the [podman API](https://docs.podman.io/en/latest/_static/api.html), not machinectl/`podman ps`. The `sudo machinectl` is also not something which cockpit can do.

This is so far apart from what cockpit does that it really doesn't fit, sorry.

**Comment by mailinglists35** [comment]
can you not talk to `/run/user/<all uids>/podman/podman.sock` just to get at least some readonly info? it looks like it has group rw permissions, so if my cockpit user is member of that group, it can talk to that user's podman api

**Comment by martinpitt** [comment]
But that is precisely what you wanted to avoid -- you went through all the trouble of creating isolated users for the containers, only to nullify that by making them all accessible by a single human user?

**Comment by mailinglists35** [comment]
> But that is precisely what you wanted to avoid -- you went through all the trouble of creating isolated users for the containers, only to nullify that by making them all accessible by a single human user?

no. I isolated the users so if one container escapes the namespace, it would limit the evil doing to it's user files instead of gaining access to all containers.

I fail to see how would giving the host human administrator access would nulify that. why would the administrator be limited in flexibility of administering the host? because that's the current state.

**Comment by martinpitt** [comment]
Ah, you want to protect the host *from* the containers, not protect the containers from the host, I see.

**Comment by mailinglists35** [comment]
@martinpitt so what do you think, could some read-only info about containers be exposed in cockpit UI by talking to those podman socket files? is this feasible for current architecture of cockpit?

**Comment by martinpitt** [comment]
Not "read-only", the podman API doesn't work that way. Possibly it could manage containers for different users, but this would be a lot of work, it far exceeds our current capacity and the scope of an issue. It doesn't really fit the design.

**Comment by mailinglists35** [comment]
> Not "read-only", the podman API doesn't work that way. Possibly it could manage containers for different users, but this would be a lot of work, it far exceeds our current capacity and the scope of an issue. It doesn't really fit the design.

i mean read only effect actions, such as listing containers, images, volumes.

come'on, it can't be _that_ complicated and out of scope!

**Comment by martinpitt** [comment]
> i mean read only effect actions

As I said, that's not how the podman API works.

You are welcome to give it a try :grin:

**Comment by mailinglists35** [comment]
> > i mean read only effect actions
> 
> As I said, that's not how the podman API works.
> 
> You are welcome to give it a try 😁

that's exactly how api works. see my previous [comment](https://github.com/cockpit-project/cockpit-podman/issues/692#issuecomment-1883108144) where I made a curl request to list containers. the output was obviously truncated but here it is again in full:

````
$ http_proxy='' ALL_PROXY='' curl --unix-socket /run/user/2804/podman/podman.sock -v 'http://d/v4.0.0/libpod/images/json' | jq
*   Trying /run/user/2804/podman/podman.sock:0...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to d (/run/user/2804/podman/podman.sock) port 80 (#0)
> GET /v4.0.0/libpod/images/json HTTP/1.1
> Host: d
> User-Agent: curl/7.76.1
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Api-Version: 1.41
< Content-Type: application/json
< Libpod-Api-Version: 4.6.1
< Server: Libpod/4.6.1 (linux)
< X-Reference-Id: 0xc0005e6008
< Date: Fri, 26 Jan 2024 13:13:57 GMT
< Transfer-Encoding: chunked
<
{ [2675 bytes data]
100  2656    0  2656    0     0   144k      0 --:--:-- --:--:-- --:--:--  144k
* Connection #0 to host d left intact
[
  {
    "Id": "a10f70804435e4db94a2d9ca9955a8cd17f7875dbac0fde4b7dddaf29be5e44f",
    "ParentId": "",
    "RepoTags": [
      "localhost/web:latest"
    ],
    "RepoDigests": [
      "localhost/web@sha256:95166b2b18d38e1e801267e612d5c493816cb32863563e30f31d452d1699cc18"
    ],
    "Created": 1705499820,
    "Size": 1423258674,
    "SharedSize": 0,
    "VirtualSize": 1423258674,
    "Labels": {
      "architecture": "x86_64",
      "build-date": "2023-12-07T06:30:54",
      "com.redhat.component": "php-81-container",
      "com.redhat.license_terms": "https://www.redhat.com/en/about/red-hat-end-user-license-agreements#UBI",
      "description": "PHP 8.1 available as container is a base platform for building and running various PHP 8.1 applications and frameworks. PHP is an HTML-embedded scripting language. PHP attempts to make it easy for developers to write dynamically generated web pages. PHP also offers built-in database integration for several commercial and non-commercial database management systems, so writing a database-enabled webpage with PHP is fairly simple. The most common use of PHP coding is probably as a replacement for CGI scripts.",
      "distribution-scope": "public",
      "help": "For more information visit https://github.com/sclorg/s2i-php-container",
      "io.buildah.version": "1.31.3",
      "io.k8s.description": "PHP 8.1 available as container is a base platform for building and running various PHP 8.1 applications and frameworks. PHP is an HTML-embedded scripting language. PHP attempts to make it easy for developers to write dynamically generated web pages. PHP also offers built-in database integration for several commercial and non-commercial database management systems, so writing a database-enabled webpage with PHP is fairly simple. The most common use of PHP coding is probably as a replacement for CGI scripts.",
      "io.k8s.display-name": "Apache 2.4 with PHP 8.1",
      "io.openshift.expose-services": "8080:http",
      "io.openshift.s2i.scripts-url": "image:///usr/libexec/s2i",
      "io.openshift.tags": "builder,php,php81,php-81",
      "io.s2i.scripts-url": "image:///usr/libexec/s2i",
      "maintainer": "SoftwareCollections.org <sclorg@redhat.com>",
      "name": "ubi9/php-81",
      "release": "43",
      "service": "web",
      "summary": "Platform for building and running PHP 8.1 applications",
      "url": "https://access.redhat.com/containers/#/registry.access.redhat.com/ubi9/php-81/images/1-43",
      "usage": "s2i build https://github.com/sclorg/s2i-php-container.git --context-dir=8.1/test/test-app ubi9/php-81 sample-server",
      "vcs-ref": "6811ef997d1c02aad25081909b79ffd6457d34f5",
      "vcs-type": "git",
      "vendor": "Red Hat, Inc.",
      "version": "1"
    },
    "Containers": 1,
    "Names": [
      "localhost/web:latest"
    ],
    "Digest": "sha256:95166b2b18d38e1e801267e612d5c493816cb32863563e30f31d452d1699cc18",
    "History": [
      "localhost/web:latest"
    ]
  }
]
````

why cockpit can NOT perform the same api request to list the running containers of some arbitrary podman socket file?

**Comment by mailinglists35** [comment]
the project makes the assumption there is only one file accessible.

https://github.com/search?q=repo%3Acockpit-project%2Fcockpit-podman%20sock&type=code

but as you could see in this conversation there are cases where multiple files are accessible, to which you can issue API calls that retrieve information.

I'm not asking to issue API calls that modify configuration in those podmans. I only need to view basic information about other containers via other socket files, which is done exactly the same as you do now with the only socket file.

**Comment by mailinglists35** [comment]
this is not always required. maybe I can enable the socket by manually starting the podman api service. it should work even when that service is not started

https://github.com/cockpit-project/cockpit-podman/blob/1d07689589fb2c3335fcd77749cf50f3791bd00a/src/app.jsx#L568

**Comment by mailinglists35** [comment]
but then again, the projects' whole philosophy is centered about logged on user's podman socket file and podman system service, ignoring the possibility that there are also other potential socket files available to issue reading information api calls to them

**Comment by martinpitt** [comment]
> > > i mean read only effect actions
> > As I said, that's not how the podman API works.
> that's exactly how api works.

That's probably a misunderstanding. I meant you cannot only do *read-only* actions -- you can either talk to the socket and do everything (including starting/stopping containers etc.) or you can't and then you can't do anything at all.

**Comment by mailinglists35** [comment]
why could you not be able to filter the list of reading apis? they're documented. on current user socket file allow all apis, on scanned socket files only issue apis that do not change the state.

**Comment by chrismaster** [comment]
Same use case here.
- one pod with some containers per rootless passwordless user
- loginctl enable-linger
- administer with machinectl and systemctl --user
- AutoUpdate=registry
I hoped for some extra security with this setup, than a rootful setup.

To check the logs and maybe in the future see if there are outdated containers, I hoped to list all containers of every user in cockpit-podman.
Perhaps there is a way to archive this

**Comment by thomasnaudin** [comment]
I wanted to achieve the exact same workflow as @chrismaster
I totally understand that there are constraints, but having the possibility to get informations about technical user account's containers via a single privileged human user account would imho really make sense, and would be a **great** enhancement.

**Comment by YamiYukiSenpai** [comment]
same here

I created my Jellyfin container on a dedicated jellyfin user account

**Comment by martinpitt** [comment]
Reopening. Our team discussed this on this week's sprint, and we found a design/solution that we can all agree to. Tracked in https://issues.redhat.com/browse/COCKPIT-1086

**Comment by martinpitt** [comment]
I posted a first draft to #2024. It still needs design and discussion, but you can already switch between users that have containers running and interact with these. Feedback appreciated! (The PR auto-builds COPRs for convenient testing on Fedora)
